morpheus/_lib/src/io/loaders/rest.cpp

```diff
@@ -17,21 +17,238 @@
+        {
+            request.set(http::field::content_type, content_type);
+            request.body() = body;
+            request.prepare_payload();
+        }
+        for (auto& x_header : x_headers)
+        {
+            request.insert(x_header.first, x_header.second);
+        }
+
+        int interval_milliseconds = 1000;
+        while (max_retry-- > 0)
+        {
+            stream.connect(endpoint_results);
+            http::write(stream, request);
+            beast::flat_buffer buffer;
+            http::read(stream, buffer, response);
+            beast::error_code ec;
+            stream.socket().shutdown(tcp::socket::shutdown_both, ec);
+            if (ec && ec != beast::errc::not_connected)
+            {
+                throw beast::system_error{ec};
+            }
+            int status = response.result_int();
+            // 503 Service Unavailable / 504 Gateway Timeout
+            if (status != 503 && status != 504)
+            {
+                break;
+            }
+            std::this_thread::sleep_for(std::chrono::milliseconds(interval_milliseconds));
+            interval_milliseconds *= 2;
+        }
+
+    } catch (std::exception const& e)
+    {
+        // TODO: need VLOG here?
+        std::cout << "Error: " << e.what() << std::endl;
+    }
+}
+
+void create_dataframe_from_http_response(http::response<http::dynamic_body>& response,
+                                         py::object& dataframe,
+                                         py::module_& mod_cudf,
+                                         const std::string& strategy)
+{
+    std::string df_json_str = beast::buffers_to_string(response.body().data());
+    boost::algorithm::trim(df_json_str);
+
+    // When calling cudf.read_json() with engine='cudf', it expects an array object as input.
+    // The workaround here is to add square brackets if the original data is not represented as an array.
+    // Submitted an issue to cudf: https://github.com/rapidsai/cudf/issues/13527.
+    if (!(df_json_str.front() == '[' && df_json_str.back() == ']'))
+    {
+        df_json_str = "[" + df_json_str + "]";
+    }
+
+    py::gil_scoped_acquire gil;
+    auto current_df = mod_cudf.attr("DataFrame")();
+    current_df      = mod_cudf.attr("read_json")(py::str(df_json_str), py::str("cudf"));
+    if (dataframe.is_none())
+    {
+        dataframe = current_df;
+    }
+    else if (strategy == "aggregate")
+    {
+        py::list args;
+        args.attr("append")(dataframe);
+        args.attr("append")(current_df);
+        dataframe = mod_cudf.attr("concat")(args);
+    }
+}
+
 std::shared_ptr<ControlMessage> RESTDataLoader::load(std::shared_ptr<ControlMessage> message, nlohmann::json task)
 {
     VLOG(30) << "Called RESTDataLoader::load()";
 
-    // TODO(Devin): Implement this
-    throw std::runtime_error("RESTDataLoader::load() not implemented yet");
+    // Aggregate dataframes for each file
+    py::gil_scoped_acquire gil;
```

Comment:
Having to do something like 'gil' and 'gil_2' is a bit clunky here. There are a couple possible workarounds. One is to adapt this and the piece at the bottom to use locally nested scope blocks, like the example I'd given above. You'll need to be a bit more careful though, since you want to fill in python objects.

Reply:
I agree. The issue here is if locally nested scope block is adapted to the first _gil_acquire_, both `mod_cudf` and `dataframe` will turn into local variable, which get out of scope at the end of the block. The workaround I could think of is to keep the explicit `gil_release` for the first `gil_acquire`, then use locally nested scope to the second `gil_acquire` to avoid something like `gil_2`